subsys/pm/device_system_managed.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(pm_device, CONFIG_PM_DEVICE_LOG_LEVEL);
+
+#define DT_PM_DEVICE_ENABLED(node_id)						\
+	COND_CODE_1(DT_PROP_OR(node_id, zephyr_pm_device_enabled, false),	\
+		(1 +), ())
+
+#define DT_PM_DEVICE_NEEDED			\
+	(DT_FOREACH_STATUS_OKAY(zephyr_power_state, DT_PM_DEVICE_ENABLED) 0)
+
+#if DT_PM_DEVICE_NEEDED
+TYPE_SECTION_START_EXTERN(const struct device *, pm_device_slots);
+
+/* Number of devices successfully suspended. */
+static size_t num_susp;
+
+bool pm_suspend_devices(void)
+{
+	const struct device *devs;
+	size_t devc;
+
+	devc = z_device_get_all_static(&devs);
+
+	num_susp = 0;
+
+	for (const struct device *dev = devs + devc - 1; dev >= devs; dev--) {
+		int ret;
+
+		/*
+		 * Ignore uninitialized devices, busy devices, wake up sources, and
+		 * devices with runtime PM enabled.
+		 */
+		if (!device_is_ready(dev) || pm_device_is_busy(dev) ||
+		    pm_device_state_is_locked(dev) ||
+		    pm_device_wakeup_is_enabled(dev) ||
+		    pm_device_runtime_is_enabled(dev)) {
+			continue;
+		}
+
+		ret = pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
+		/* ignore devices not supporting or already at the given state */
+		if ((ret == -ENOSYS) || (ret == -ENOTSUP) || (ret == -EALREADY)) {
+			continue;
+		} else if (ret < 0) {
+			LOG_ERR("Device %s did not enter %s state (%d)",
+				dev->name,
+				pm_device_state_str(PM_DEVICE_STATE_SUSPENDED),
+				ret);
+			return false;
+		}
+
+		TYPE_SECTION_START(pm_device_slots)[num_susp] = dev;
```

Comment:
Why is this updating an array in RAM instead of setting a flag on the device PM struct?

Reply:
The idea was to have a list of only suspended devices to resume (in reverse order) when the system is resumed avoiding need to check all devices.
Honestly, thinking now it seems that paying this price is better than having this pre-allocated array in ram ...

This logic was not introduced in these changes though, we can review it separately I believe.